File: Data Structures & Algorithms/buy-and-sell-crypto/submission-2.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        max_profit = 0

        for i in range(len(prices)):
            buy_price = prices[i]

            for j in range(i+1, len(prices)):
                sell = prices[j]

                max_profit = max(max_profit, sell - buy_price)
            
        return max_profit



"""
prices = [10,1,5,6,7,1]
             i
max = 0

buy = 1
sell = 5

0 > 0

"""

File: Data Structures & Algorithms/buy-and-sell-crypto-with-cooldown/submission-2.java
class Solution {
    public int maxProfit(int[] prices) {
        return f(0, 0, prices);
    }

    public int f(int idx, int canSell, int[] prices) {
        if(idx >= prices.length) {
            return 0;
        }

        int profit = 0;

        if(canSell == 0) {
            int buy = - prices[idx] + f(idx + 1, 1, prices);
            int skipBuy = 0 + f(idx + 1, 0, prices);

            profit = Math.max(buy, skipBuy);
        }

        if(canSell == 1) {
            int sell = + prices[idx] + f(idx + 2, 0, prices);
            int skipSell = 0 + f(idx + 1, 1, prices);

            profit = Math.max(sell, skipSell);
        }

        return profit;
    }
}
